Make WebCodecsEncodedVideoChunk serializable

youennf opened the pull request:
#### 843735d1dc8b0339e5764363f3f540dd4ec36874
<pre>
Make WebCodecsEncodedVideoChunk serializable
<a href="https://bugs.webkit.org/show_bug.cgi?id=246610">https://bugs.webkit.org/show_bug.cgi?id=246610</a>
rdar://problem/101239648

Reviewed by Eric Carlson.

Introduce WebCodecsEncodedVideoChunkStorage as thread safe ref counted and have WebCodecsEncodedVideoChunk own a WebCodecsEncodedVideoChunkStorage.
Use WebCodecsEncodedVideoChunkStorage to support serialization.
Either the data will be encoded as IPC or the WebCodecsEncodedVideoChunkStorage pointer will be retrieved from a the serialized value vector.
Update RTCDataChannelTransferTag description to match what we are doing (send an ID).

* LayoutTests/imported/w3c/web-platform-tests/webcodecs/chunk-serialization.any-expected.txt:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/webcodecs/WebCodecsEncodedVideoChunk.cpp:
(WebCore::WebCodecsEncodedVideoChunk::WebCodecsEncodedVideoChunk):
(WebCore::WebCodecsEncodedVideoChunk::copyTo):
* Source/WebCore/Modules/webcodecs/WebCodecsEncodedVideoChunk.h:
(WebCore::WebCodecsEncodedVideoChunkStorage::create):
(WebCore::WebCodecsEncodedVideoChunkStorage::data const):
(WebCore::WebCodecsEncodedVideoChunkStorage::memoryCost const):
(WebCore::WebCodecsEncodedVideoChunk::create):
(WebCore::WebCodecsEncodedVideoChunk::type const):
(WebCore::WebCodecsEncodedVideoChunk::timestamp const):
(WebCore::WebCodecsEncodedVideoChunk::duration const):
(WebCore::WebCodecsEncodedVideoChunk::byteLength const):
(WebCore::WebCodecsEncodedVideoChunk::data const):
(WebCore::WebCodecsEncodedVideoChunk::storage):
(WebCore::WebCodecsEncodedVideoChunkStorage::WebCodecsEncodedVideoChunkStorage):
(WebCore::WebCodecsEncodedVideoChunk::WebCodecsEncodedVideoChunk):
* Source/WebCore/Modules/webcodecs/WebCodecsEncodedVideoChunkData.h: Added.
* Source/WebCore/Modules/webcodecs/WebCodecsEncodedVideoChunkType.h:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/SerializedScriptValue.cpp:
(WebCore::CloneSerializer::serialize):
(WebCore::CloneSerializer::CloneSerializer):
(WebCore::CloneSerializer::dumpWebCodecsEncodedVideoChunk):
(WebCore::CloneSerializer::dumpIfTerminal):
(WebCore::CloneDeserializer::deserialize):
(WebCore::CloneDeserializer::CloneDeserializer):
(WebCore::CloneDeserializer::readWebCodecsEncodedVideoChunk):
(WebCore::CloneDeserializer::readTerminal):
(WebCore::SerializedScriptValue::SerializedScriptValue):
(WebCore::SerializedScriptValue::computeMemoryCost const):
(WebCore::SerializedScriptValue::create):
(WebCore::SerializedScriptValue::deserialize):
* Source/WebCore/bindings/js/SerializedScriptValue.h:
(WebCore::SerializedScriptValue::encode const):
(WebCore::SerializedScriptValue::decode):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: <a href="https://commits.webkit.org/255718@main">https://commits.webkit.org/255718@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f7c6294db98e0aca271d7e165bb4752654a7ba6c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/93422 "7 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/2618 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/24066 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/103094 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/163427 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/97420 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/2628 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/30920 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/85788 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/99199 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/99085 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/1850 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/79873 "Built successfully") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/28893 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/83763 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/83513 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/71852 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/37303 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/17380 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/35133 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/18630 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3955 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/39006 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/41156 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/40942 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/37853 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

eric-carlson commented on Source/WebCore/bindings/js/SerializedScriptValue.cpp:
```diff
@@ -417,8 +422,9 @@ static const unsigned StringDataIs8BitFlag = 0x80000000;
  *    | OffscreenCanvasTransferTag <value:uint32_t>
  *    | WasmMemoryTag <value:uint32_t>
- *    | RTCDataChannelTransferTag <processIdentifier:uint64_t><rtcDataChannelIdentifier:uint64_t><label:String>
+ *    | RTCDataChannelTransferTag <identifier:uint32_t>
```
If this change is intentional, you should probably describe it in the commit log.

youennf replied:
Will do.

eric-carlson reviewed (approved):
r=me

youennf committed:
Make WebCodecsEncodedVideoChunk serializable
https://bugs.webkit.org/show_bug.cgi?id=246610
rdar://problem/101239648

Reviewed by Eric Carlson.

Introduce WebCodecsEncodedVideoChunkStorage as thread safe ref counted and have WebCodecsEncodedVideoChunk own a WebCodecsEncodedVideoChunkStorage.
Use WebCodecsEncodedVideoChunkStorage to support serialization.
Either the data will be encoded as IPC or the WebCodecsEncodedVideoChunkStorage pointer will be retrieved from a the serialized value vector.
Update RTCDataChannelTransferTag description to match what we are doing (send an ID).

* LayoutTests/imported/w3c/web-platform-tests/webcodecs/chunk-serialization.any-expected.txt:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/webcodecs/WebCodecsEncodedVideoChunk.cpp:
(WebCore::WebCodecsEncodedVideoChunk::WebCodecsEncodedVideoChunk):
(WebCore::WebCodecsEncodedVideoChunk::copyTo):
* Source/WebCore/Modules/webcodecs/WebCodecsEncodedVideoChunk.h:
(WebCore::WebCodecsEncodedVideoChunkStorage::create):
(WebCore::WebCodecsEncodedVideoChunkStorage::data const):
(WebCore::WebCodecsEncodedVideoChunkStorage::memoryCost const):
(WebCore::WebCodecsEncodedVideoChunk::create):
(WebCore::WebCodecsEncodedVideoChunk::type const):
(WebCore::WebCodecsEncodedVideoChunk::timestamp const):
(WebCore::WebCodecsEncodedVideoChunk::duration const):
(WebCore::WebCodecsEncodedVideoChunk::byteLength const):
(WebCore::WebCodecsEncodedVideoChunk::data const):
(WebCore::WebCodecsEncodedVideoChunk::storage):
(WebCore::WebCodecsEncodedVideoChunkStorage::WebCodecsEncodedVideoChunkStorage):
(WebCore::WebCodecsEncodedVideoChunk::WebCodecsEncodedVideoChunk):
* Source/WebCore/Modules/webcodecs/WebCodecsEncodedVideoChunkData.h: Added.
* Source/WebCore/Modules/webcodecs/WebCodecsEncodedVideoChunkType.h:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/SerializedScriptValue.cpp:
(WebCore::CloneSerializer::serialize):
(WebCore::CloneSerializer::CloneSerializer):
(WebCore::CloneSerializer::dumpWebCodecsEncodedVideoChunk):
(WebCore::CloneSerializer::dumpIfTerminal):
(WebCore::CloneDeserializer::deserialize):
(WebCore::CloneDeserializer::CloneDeserializer):
(WebCore::CloneDeserializer::readWebCodecsEncodedVideoChunk):
(WebCore::CloneDeserializer::readTerminal):
(WebCore::SerializedScriptValue::SerializedScriptValue):
(WebCore::SerializedScriptValue::computeMemoryCost const):
(WebCore::SerializedScriptValue::create):
(WebCore::SerializedScriptValue::deserialize):
* Source/WebCore/bindings/js/SerializedScriptValue.h:
(WebCore::SerializedScriptValue::encode const):
(WebCore::SerializedScriptValue::decode):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: https://commits.webkit.org/255718@main
